### ISSUE TITLE
Unable to run largish test suite due to timeout/hang

### ISSUE THREAD

**Issue by lassedamgaard** [dart-lang/test#317]
I have a suite of 800+ browser based tests spread across some ~80 files. 
Using either content-shell or dartium the test run times out after around 1 minute and 40 seconds. In content-shell the runner just seems to hang. In dartium I get a connection closed error. I'm pretty sure it's not a bad test that causes this because the tests that are not run on dartium can run on content shell. Similarly I can skip some groups longer up the chain to reduce the total number of tests and thus make the last part of the run succeed. But I cannot run everything. 

Using test 0.12.4 and dart 1.10 on OSX and concurrency 1. With higher concurrency the timeout occurs earlier. 

I have no idea how to debug this.

Example run in content shell (last line hangs forever):

```
01:37 +686 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils deploying a revision older than some of the target deployment revisions is a warning
01:37 +687 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils downgrade warning is skipped if commit query fails
01:37 +688 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils downgrade warning is skipped if commit query fails
01:37 +689 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils deploying a new build when a reference build exists is a warning
01:37 +690 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils deploying a different build than the reference build is a warning
01:37 +691 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils deploying a different build than some of multiple reference builds is a warning
01:37 +692 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils deploying a different build than any of multiple reference builds is a warning
01:37 +693 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils deploying a different build than some of multiple reference builds is a warning
01:37 +694 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils deploying a different ref than all multiple reference builds is a warning
01:37 +695 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils reference deployment check proceeds if only some reference builds cannot be retrieved
01:37 +696 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils reference deployment check is aborted if no reference builds can be retrieved
01:37 +697 ~4: test/test_upgrade_dialog_utils.dart: upgrade_utils reference deployment check is aborted if target build cannot be retrieved
01:37 +698 ~4: test/test_upgrade_options.dart: upgrade_options view properties are correctly initialized from deployment data and can be changed
01:38 +699 ~4: loading test/test_workflow_diagram.dart
01:38 +699 ~4: test/test_upgrade_options.dart: upgrade_options batch size, health checks and canary options are not used when upgrading dispatch
01:39 +700 ~4: test/test_upgrade_options.dart: upgrade_options enabling rolling dispatch upgrade sets batch size param
```

And in dartium:
```
01:31 +642 ~4: test/test_uconfig_view.dart: uconfig_view full view when reinit returns same branch then a notification is shown
01:31 +643 ~4: test/test_uconfig_view.dart: uconfig_view full view when reinit returns a new branch then the new branch is selected
01:31 +644 ~4: test/test_uconfig_view.dart: uconfig_view full view when reload service defs is clicked, new definitions are loaded and inserted into ui
01:31 +645 ~4: test/test_uconfig_view.dart: uconfig_view full view when merging to master gives a conflict then a dialog is shown with instructions
01:31 +646 ~4: test/test_uconfig_view.dart: uconfig_view full view importing from flipr sets new branch if one is returned from service
01:31 +647 ~4: test/test_uconfig_view.dart: uconfig_view full view clicking the history icon opens the history dialog
01:32 +648 ~4: test/test_uconfig_view.dart: uconfig_view full view no history icon is shown for a new field
01:32 +649 ~4: test/test_uconfig_view.dart: uconfig_view full view no delete icon is shown for a placeholder field
01:32 +650 ~4: test/test_uconfig_view.dart: uconfig_view full view placeholder is not shown for dict field when field has a child value
01:32 +651 ~4: test/test_uconfig_view.dart: uconfig_view full view list type items have add icon shown
01:32 +652 ~4: test/test_uconfig_view.dart: uconfig_view full view clicking the add item icon adds empty value with next index
01:32 +653 ~4: test/test_update_monitor.dart: update_monitor object registers with query service and unregisters when index changes after dispatching a notification
01:37 +654 ~4: test/test_upgrade_dialog.dart: upgrade_dialog upgrade can be started from a new build
01:38 +654 ~4 -1: loading test/test_upgrade_dialog_utils.dart
  Failed to load "test/test_upgrade_dialog_utils.dart": Connection closed before test suite loaded.
01:38 +654 ~4 -2: loading test/test_upgrade_options.dart
  Bad state: The underlying channel is closed.
  dart:isolate  _RawReceivePortImpl._handleMessage
  
01:38 +654 ~4 -3: loading test/test_workflow_diagram.dart
  Bad state: The underlying channel is closed.
  dart:isolate  _RawReceivePortImpl._handleMessage
  
01:38 +654 ~4 -4: loading test/test_workflow_service.dart
  Bad state: The underlying channel is closed.
  dart:isolate  _RawReceivePortImpl._handleMessage
  
```

**Comment by nex3** [comment]
When you say "timeout", do you mean that the test runner is reporting a timeout and closing, or do you mean that the test runner hangs and doesn't print any more progress?

Can you see anything unusual if you run content shell with `--pause-after-load` and connect to the Observatory? What if you use a local copy of `test`, comment out [this line](https://github.com/dart-lang/test/blob/master/lib/src/runner/browser/browser.dart#L116), and watch Dartium's console to see if it's printing anything unusual?

Is there any way you can send me your test suite so I can try to reproduce this locally?

**Comment by lassedamgaard** [comment]
@nex3 The terminal snippets in the description is simply that last output from the testrunner and I have to kill the process. So when running against content-shell it just hangs after printing the result of the last test that ran and in dartium it hangs after printing "Failed to load blabla... underlying connection closed... etc" that I pasted above. So 'timeout' was not really the most descriptive word.

I will try your suggestions and let you know what I find.

**Comment by lassedamgaard** [comment]
OK I should have dug just a little deeper before filing this. What actually happens is not a timeout at all but simply that dartium/content-shell crashes after eating up about 2.5 GB of memory. Memory usage steadily climbs from a few hundred MB until the process dies at 2.5 GB. 

This is the output from ```top -pid <PID> -l 0``` just before and after ```Chromium Helper``` dies:
```
PID   COMMAND         %CPU  TIME     #TH  #WQ #PORTS MEM    PURG  CMPRS PGRP PPID STATE    BOOSTS  %CPU_ME %CPU_OTHRS UID FAULTS   COW   MSGSENT MSGRECV SYSBSD   SYSMACH CSW     PAGEINS IDLEW  POWER USER  #MREGS RPRVT VPRVT VSIZE KPRVT KSHRD
4913- Chromium Helper 12.6  01:09.29 17   1   127    2454M+ 260K  0B    4906 4910 stuck    *0[27]  0.00000 0.00000    501 1019064+ 9933  32280+  22871+  1955384+ 284557+ 210203+ 1400    14998+ 35.6  lasse N/A    N/A   N/A   N/A   N/A   N/A  
Processes: 245 total, 2 running, 11 stuck, 232 sleeping, 1365 threads 
2015/08/13 10:18:01
Load Avg: 4.17, 3.12, 3.25 
CPU usage: 7.22% user, 4.33% sys, 88.43% idle 
SharedLibs: 16M resident, 14M data, 0B linkedit.
MemRegions: 79169 total, 4880M resident, 73M private, 1512M shared.
PhysMem: 11G used (1877M wired), 4826M unused.
VM: 634G vsize, 1065M framework vsize, 4400089(0) swapins, 4875340(0) swapouts.
Networks: packets: 9296936/9194M in, 8158653/5228M out.
Disks: 2415808/42G read, 2885295/75G written.

PID   COMMAND         %CPU  TIME     #TH  #WQ #PORTS MEM    PURG  CMPRS PGRP PPID STATE    BOOSTS  %CPU_ME %CPU_OTHRS UID FAULTS   COW   MSGSENT MSGRECV SYSBSD   SYSMACH CSW     PAGEINS IDLEW  POWER USER  #MREGS RPRVT VPRVT VSIZE KPRVT KSHRD
Processes: 243 total, 2 running, 11 stuck, 230 sleeping, 1362 threads 
2015/08/13 10:18:02
Load Avg: 4.23, 3.15, 3.26 
CPU usage: 8.9% user, 5.79% sys, 86.11% idle 
SharedLibs: 16M resident, 14M data, 0B linkedit.
MemRegions: 78154 total, 4846M resident, 73M private, 1512M shared.
PhysMem: 10G used (1858M wired), 6033M unused.
VM: 632G vsize, 1065M framework vsize, 4400089(0) swapins, 4875340(0) swapouts.
Networks: packets: 9296982/9194M in, 8158698/5228M out.
Disks: 2415817/42G read, 2885398/75G written.
```

I see the following error printed repeatedly (looks like it's for each test file loaded) in devtools:

![dartium test](https://cloud.githubusercontent.com/assets/6475621/9245956/6ff4616e-41a8-11e5-86ec-ee09a060f25d.png)

I tried running with ```--pause-after-load``` and check out observatory but this does not really work: When running the first test file I can open observatory just fine and I don't see anything obviously suspicious, but then when the next file is loaded the test run just hangs again.

I also tried removing the process kill in browser.dart and this made no difference as far as I can tell.

Could this be related to the fact that there are many test files being loaded and run instead of everything being run from a single entry point?

**Comment by lassedamgaard** [comment]
Seems that this is indeed related to the number of individual test files being loaded. If I generate a single runner file that just imports all 80 test files and just runs their main() method the memory usage stays around ~200 MB and test suite runs to completion. 
```dart
import 'test_foo.dart' as test_foo;


main() {
  test_abort_workflow_dialog.main();
  test_aggregator_api_mapper.main();

```

**Comment by nex3** [comment]
It looks like the host frame is failing to remove iframes as they become unused, which explains the memory leak. I can reproduce that; I'll look into it.

**Comment by lassedamgaard** [comment]
I can confirm that the memory leak is fixed in 0.12.4.+3
Thanks for the quick fix  :+1:
